### ISSUE TITLE
Weird changes in API docs

### ISSUE THREAD

**Issue by psychon** [awesomeWM/awesome#834]
https://github.com/awesomeWM/apidoc/commit/e3619cc30000a3462fb87575245fc62524bd3285#diff-ac83493a098e53ca8086890f2845a7d5L3711
https://github.com/awesomeWM/apidoc/compare/dba588b20c...e3619cc300#diff-7d61cae60cd4bcbe5e620947dc757a9aL796

This changes the target of a link in a a build that didn't touch anything in awful (it's the build of my recent naughty PR). The actual link target (`<a id="foo">`) is not modified. So either before or after this change, the link was broken.

What is ldoc doing here? Do we have docs with broken links?

**Comment by blueyed** [comment]
The first change is:

```diff
-         <a href="../classes/client.html#awful.client.get_transient_for_matching">client.get_transient_for_matching</a>
+         <a href="../classes/client.html#client:get_transient_for_matching">client.get_transient_for_matching</a>
```

I've seen some changes like this before, and thought it was because of some
(ldoc?) update during the build.

Both links seem to work after all:

http://awesomewm.github.io/apidoc/classes/client.html#awful.client.get_transient_for_matching
http://awesomewm.github.io/apidoc/classes/client.html#client:get_transient_for_matching

But one appears to be for the (shortened) instance, and the other for the longer one.
But after all those should not get linked to `client`, but `awful.client`.

/cc @Elv13

**Comment by Elv13** [comment]
`awful.client` is "gone" (from ldoc), `awful.client.get_transient_for_matching is deprecated` `client:get_transient_for_matching()` is the non deprecated function. Maybe ldoc use a hashtable and the first element matching happen to change. In that case it would need to be fixed.

**Comment by blueyed** [comment]
Ok, got it.

The issue seems to be though that there are too many changes (and not just the ones from the update/merge).

Relevant log: https://travis-ci.org/awesomeWM/awesome/jobs/123780619#L3360
> 5 files changed, 11 insertions(+), 11 deletions(-)

There are also changes for https://github.com/awesomeWM/awesome/compare/3fe85a8f050a...8cfdc0c59334: https://github.com/awesomeWM/apidoc/commit/ca56189e5ab1696595e18253632f9a7cc9d236c6 ?!

I guess it's time for some review of https://github.com/awesomeWM/awesome/blob/master/build-utils/travis-apidoc.sh.. :)

**Comment by blueyed** [comment]
It's not related to the apidoc build script, but luadoc itself, like @Elv13 suggested.

Try the following (e.g. on 5f8dd56):

```sh
% for i in {1..5}; do make ldoc >/dev/null; cp -a .build-*/doc doc-$i; [ "$i" = 1 ] && continue; diff -Nur doc-1 doc-$i -I "Last updated" -I "<strong>Release</strong>:" \
  -I "<h2>API documentation for awesome, a highly configurable X window manager (version .*)\.</h2>" \
  -x .git | md5sum; done
```

The md5 sum should be the same always, but is not.

**Comment by psychon** [comment]
Daniel, did you try running that without "| md5sum"? Some random example I get is https://gist.github.com/psychon/a387bc57722e5d990bef9d2279082763 and here https://gist.github.com/psychon/a387bc57722e5d990bef9d2279082763#file-diff-patch-L109-L111 makes me wonder what the heck is going on. It's not even a small difference, but something is off by about four pixels.

(And of course changes like the following are always fun)
```diff
-    Default handler for tiled clients request::geometry with the <a href="../libraries/mouse.html#awful.mouse.client.move">mouse.move</a>
+    Default handler for tiled clients request::geometry with the <a href="../libraries/mouse.html#awful.mouse.wibox.move">mouse.move</a>
```

**Comment by blueyed** [comment]
Yeah, the md5sum was meant to just show that the differences are changing all the time.

@Elv13 
Any idea about why the svg changes randomly?

**Comment by Elv13** [comment]
Any examples of the changes (I know, I am lazy)?

A theory is that given the SVG use long floating point numbers, Cairo may be compiled with `-ffast-math` to improve the speed. This cause some optimizations to be made. Some of those are context sensitive and non-deterministic. The answer is "very close", but not "exactly precise". Given it cause an "error" smaller than a pixel, those optimizations are mostly harmless. Other than that, SVG may contain random numbers identifiers when creating "classes" that can be re-used (such as patterns). I am not sure if cairo use these.

(unrelated SVG is the right choice here. It supports HiDPI and mobile viewing without artifacts, the system can also output "native" PNGs).

**Comment by blueyed** [comment]
@Elv13 
One example from @psychon's diff:

```diff
-<path style="fill:none;stroke-width:1.5;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(38.039216%,50.588235%,100%);stroke-opacity:1;stroke-miterlimit:10;" d="M 0 5 C 0 2.238281 2.238281 0 5 0 L 31 0 C 33.761719 0 36 2.238281 36 5 L 36 16 C 36 18.761719 33.761719 21 31 21 L 5 21 C 2.238281 21 0 18.761719 0 16 Z M 0 5 " transform="matrix(1,0,0,1,22,10)"/>
+<path style="fill:none;stroke-width:1.5;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(38.039216%,50.588235%,100%);stroke-opacity:1;stroke-miterlimit:10;" d="M 0 5 C 0 2.238281 2.238281 0 5 0 L 27 0 C 29.761719 0 32 2.238281 32 5 L 32 6 C 32 8.761719 29.761719 11 27 11 L 5 11 C 2.238281 11 0 8.761719 0 6 Z M 0 5 " transform="matrix(1,0,0,1,10,16)"/>
```

**Comment by Elv13** [comment]
Umm... That's not small at all, but that's also only one SVG out of ~40. Does it happen in all doc updates?

**Comment by blueyed** [comment]
This one seems to be affected always recently, others not.

And there's only the `ldoc` related changes, which seem to be different / another issue?!

**Comment by blueyed** [comment]
@Elv13 @psychon 
Would be nice to get this solved, so the relevant changes from the bot's commits are useful finally.

**Comment by Elv13** [comment]
I don't think it can be solved permanently, see my comment in #897. I suggest closing this.

**Comment by psychon** [comment]
> One example from @psychon's diff:
> ``` diff
> -<path style="fill:none;stroke-width:1.5;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(38.039216%,50.588235%,100%);stroke-opacity:1;stroke-miterlimit:10;" d="M 0 5 C 0 2.238281 2.238281 0 5 0 L 31 0 C 33.761719 0 36 2.238281 36 5 L 36 16 C 36 18.761719 33.761719 21 31 21 L 5 21 C 2.238281 21 0 18.761719 0 16 Z M 0 5 " transform="matrix(1,0,0,1,22,10)"/>
> +<path style="fill:none;stroke-width:1.5;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(38.039216%,50.588235%,100%);stroke-opacity:1;stroke-miterlimit:10;" d="M 0 5 C 0 2.238281 2.238281 0 5 0 L 27 0 C 29.761719 0 32 2.238281 32 5 L 32 6 C 32 8.761719 29.761719 11 27 11 L 5 11 C 2.238281 11 0 8.761719 0 6 Z M 0 5 " transform="matrix(1,0,0,1,10,16)"/>
> ```

I just came across this again and I can conclude that the following is wrong:

> A theory is that given the SVG use long floating point numbers, Cairo may be compiled with -ffast-math to improve the speed.

Here is a diff of such a change (dunno if the link works): https://github.com/awesomeWM/apidoc/commit/9963bf794e719c6937d3bd342ec78cc66b1cdb3b?short_path=007858f#diff-007858fd08ac9caaefd4384588837b53

And here are the two images:
https://github.com/awesomeWM/apidoc/blob/9963bf794e719c6937d3bd342ec78cc66b1cdb3b/images/AUTOGEN_wibox_widget_piechart_label.svg
https://github.com/awesomeWM/apidoc/blob/21583d2133edab6adf6dfce7fb5105d9e16d6233/images/AUTOGEN_wibox_widget_piechart_label.svg

As you can see, the order in which the piechart draws things changes non-deterministically. And that's a serious bug. This means that a redraw might not actually work!
Imagine if only half the widget is redrawn, but the widget now decides to change the order of things. The result will be broken.

I think we should take this issue ("Weird changes in the API doc") much more serious. So far, all of the issues turned out to be bugs (with the exception of `tests/examples/text/gears/object/signal.lua` printing an object address, but still, that one should be fixed).

**Comment by psychon** [comment]
The same issue is present in `tests/examples/awful/placement/resize_to_mouse.lua`. It uses `pairs()` to determine the order in which things will appear.
![changes2](https://cloud.githubusercontent.com/assets/89482/20834759/36ac3942-b897-11e6-8026-27934aff4c7b.png)

**Comment by blueyed** [comment]
> I think we should take this issue much more serious.

I agree - even without knowing that it reveals actual bugs.. ;)

**Comment by blueyed** [comment]
I wish it was fixed, but probably is not.

**Comment by blueyed** [comment]
This ldoc patch appears to help for the mix between `{tag,beautiful}.master_fill_policy` etc: https://github.com/stevedonovan/LDoc/pull/257.

**Comment by blueyed** [comment]
Progress!  I've made it to have a stable ldoc outcome - but need to clean it up tomorrow.

**Comment by psychon** [comment]
Forgot to add a link: #1703
